advanced/src/hashTable.js

```diff
@@ -0,0 +1,258 @@
+        return currentNode.value
+      } else {
+        currentNode = currentNode.next
+      }
+    } return -1
+  }
+
+//see if linked list contains key
+  contains(key) {
+    var currentNode = this.head
+    if (!currentNode) {
+      return null
+    }
+    while (currentNode) {
+      if (currentNode.key === key){
+        return true
+      } else {
+        currentNode = currentNode.next
+      }
+    } return false
+  }
+
+
+//finds the node whose next value is the target node
+  findByNext(key) {
+    let currentNode = this.head
+
+    if(!currentNode) {
+      return null
+    } else if (!currentNode.next && currentNode.key === key){
+      return clear()
+    } else {
+      while(currentNode.next){
+        if(currentNode.next.key === key){
+          return currentNode
+        } else {
+          currentNode = currentNode.next
+        }
+      } return null
+    }
+  }
+
+//cleaner remove using findByNext
+  removeWithFindByNext(key) {
+    let currentNode = this.head
+
+    const findByNext = function(key) {
+      if(!currentNode) {
+        return null
+      } else if (!currentNode.next && currentNode.key === key){
+        return clear()
+      } else {
+        while(currentNode.next){
+          if(currentNode.next.key === key){
+            return currentNode
+          } else {
+            currentNode = currentNode.next
+          }
+        } return null
+      }
+    }
+
+    let foundNode = findByNext(key)
+    if (foundNode === null){
+      return null
+    } else {
+      foundNode.next = foundNode.next.next
+      return this.length--
+    }
+  }
+
+//brute force
+  removeIt(key) {
+    let currentNode = this.head
+
+    if(!currentNode){
+      return null
+    } else {
+      while(currentNode){
+        if(currentNode.key === key){
+          if(!currentNode.next){
+            this.head = null
+            return this.length--
+          } else {
+            currentNode = currentNode.next
+            this.length--
+            return this.head = currentNode
+          }
+        } else if (currentNode.next.key === key){
+          currentNode.next = currentNode.next.next
+          return this.length--
+        } else if (currentNode.next !== null){
+          currentNode = currentNode.next
+        } else {
+          return null
+        }
+      }
+    }
+  }
+
+}
+
+export default class HashTable {
+  constructor() {
+    this.length = 0
+    this.table = {}
+  }
+
+  //ht.iterate((k, v) => console.log(`${k}: ${v}`))
+    // takes a callback function and passes it each key and value in sequence.
+  iterate( callback( k, v ) ){
+    let table = this.table
+    let arrayOfKeys = Object.keys(table)
+
+    for( let i = 0; i < arrayOfKeys.length; i++ ){
+      callback( arrayOfKeys[i], table[arrayOfKeys[i]] )
+    }
+  }
+
+  //ht.hash("name")
+    // generates a hash for the key "name"
+  hash( str ) {
+    const inputString = str
+    const strToArray = inputString.toLowerCase().split('')
```

Comment:
I dont think you need `strToArray`. both `inputString` and `str` should support `.length` and `[0]` syntax so you can skip this step